### ISSUE TITLE
Possible race condition when initActivity throws an Exception

### ISSUE THREAD

**Issue by activeeon-bot** [ow2-proactive/programming#501]
<a href="https://jira.activeeon.com/browse/PROACTIVE-620" title="PROACTIVE-620">Original issue</a> created by <a href="mailto:clement.mathieu_AT_activeeon.com">Clement Mathieu</a> on 04, Mar 2009 at 17:35 PM - PROACTIVE-620

<hr />

When the initActivity throws an Exception, a race condition can occurs. It results in a client waiting forever blocked on FutureProxy.waitFor()

1- AO creation
2- Client calls a method on the AO
3- Client waits for the reply
3- AO runs initActivity
4- AO terminates

A possible fix for this issue is to open the threadstore only when initActivity has been run.

**Comment by activeeon-bot** [comment]
Attachment:
- [0001-Fixed-PROACTIVE-620.patch](https://github.com/ow2-proactive/backup-attachments-jira/blob/gh-pages/proactive/620/0001-Fixed-PROACTIVE-620.patch)

**Comment by activeeon-bot** [comment]
<a href="https://jira.activeeon.com/browse/PROACTIVE-620?focusedCommentId=11260">Original comment</a> posted by <a href="mailto:clement.mathieu_AT_activeeon.com">Clement Mathieu</a> on 05, Mar 2009 at 13:30 PM

<hr />

I have tried to fix this issue by modifying the threadstore's default state. Threadstores are now closed by default. ActiveBody and HalfBody must open the ThreadStore when the body is ready.

Unfortunately a such change modify the ProActive behaviour. Indeed, it was possible to call getStubOnThis() from the initActivity() method and then to perform local calls through the stub. If the threadstore is closed when the initActivity()  is run, a deadlock occurs.

It is obvious that what don't apply this patch in the 4.0.x branch. It is a major functional change.

What about ProActive 4.1.0 ?
1- Don't change anything
2- Patch the threadstore and document the new limitation
3- Find another way to fix this bug that is backward compatible 

IMHO we must decide if the body is ready when initActivity() is run or not. It is currently unspecified.

**Comment by activeeon-bot** [comment]
<a href="https://jira.activeeon.com/browse/PROACTIVE-620?focusedCommentId=11261">Original comment</a> posted by <a href="mailto:clement.mathieu_AT_activeeon.com">Clement Mathieu</a> on 05, Mar 2009 at 13:31 PM

<hr />

Unless a backward compatible patch is found, it will not be applied to the 4.0.x branch.

**Comment by activeeon-bot** [comment]
<a href="https://jira.activeeon.com/browse/PROACTIVE-620?focusedCommentId=11268">Original comment</a> posted by <a href="mailto:fabien.viale_AT_activeeon.com">Fabien Viale</a> on 05, Mar 2009 at 15:13 PM

<hr />

In the MasterWorker I'm in the case where it's quite convenient to use the stub and perform a call through it inside the InitActivity method of the class AOWorker 

This call is a oneway call and is expected to be the fist request executed during the worker life cycle, after the init activity, the method will connect to the master and get tasks which need to be run. So the worker is reactive and ready to be used as soon as it's created.

Forbidding this behaviour would mean that I would need to force this first call in a RunActivity method for AOWorker, this would mean creating myself the MethodCall, Request, finding the approprate Exception context, etc...

I believe it's possible, but I might have to write quite dirty code in order to reach that.

**Comment by activeeon-bot** [comment]
<a href="https://jira.activeeon.com/browse/PROACTIVE-620?focusedCommentId=13001">Original comment</a> posted by <a href="mailto:clement.mathieu_AT_activeeon.com">Clement Mathieu</a> on 01, Oct 2010 at 17:40 PM

<hr />

Fix. Functional change, careful review required

**Comment by activeeon-bot** [comment]
<a href="https://jira.activeeon.com/browse/PROACTIVE-620?focusedCommentId=18616">Original comment</a> posted by <a href="mailto:youri.bonnaffe_AT_activeeon.com">Youri Bonnaffe</a> on 12, Dec 2014 at 13:00 PM

<hr />

Marking old issues as won't fix, either because we don't support it anymore, or it is needed or it has been there for too long.

Feel free to reopen if needed.
